Fix for Microphone Streaming Inconsistencies (Booleans/Error Codes for Initialisation, Starting Streams and Recording)

DarcyWilson-MU opened the pull request:
## Overview
When using the MRTK2 Examples, specifically the Audio Demo related to the **_Windows Microphone Stream_** system, there would be issues when trying to switch the state of the microphone. A modification of the demo was made to initialise the stream using the Room stream type, record thirty seconds of audio, uninitialise it and then reinitialise it with the High Quality Voice stream type. However, the state of the microphone, whether it was initialised, recording or streaming, would never change and would just skip over the checks to see if they were still running or not. This would cause some errors where it would believe it hadn't started streaming or been initialised, and made it so attempting to switch between stream types would fail as it had never been uninitialised, or recordings would never start.

This addresses those issues and adds more error codes to both the scripts and plugin files to make it easier to spot issues. The reasoning being that the `SUCCESS` code, where correct as it is working, isn't too descriptive and could possibly lead to 

## Changes
- Added lines to set state Booleans [[d40a22a]](https://github.com/microsoft/MixedRealityToolkit/commit/d40a22a0b5fe61ccbbdabba9db583e81ccef1630)
- Changed the return type of methods to better reflect current stream state [[31f3323]](https://github.com/microsoft/MixedRealityToolkit/commit/31f3323b65410da66ee37122f0497ba82e3cb101)
- Added new stream error codes [[319f2b6]](https://github.com/microsoft/MixedRealityToolkit/commit/319f2b60e19bf243625522b0e0e8d8c8bb6f97de)

DarcyWilson-MU committed:
Update WindowsMicrophoneStream.cs

NOTES:
- Added lines to set the state of the various Booleans used to keep track of the initialisation, streaming and recording of the microphone stream. The way it was set up before, in most instances these Booleans would not change, and portions of the code would never fire off as they relied on them to do so.

DarcyWilson-MU committed:
Update WindowsMicrophoneStream.cs

NOTES:
- Changed the `WindowsMicrophoneStreamErrorCode` return types from `Success` to `AlreadyRunning` for `Initialize()` and `AlreadyRecording` for `StartRecording()` if the Booleans tracking those states are already true. Thought it to be more descript than setting everything to `Success` which is technically true as it is working, but it doesn't let the developer know that these states were already running and may cause them to add code where not needed.

DarcyWilson-MU committed:
Added New Microphone Stream Error Codes

NOTES:
- Added new stream error codes to better indicate to the developer that a microphone stream state has or has not already started. Similar reasoning to the previous push, setting everything as `Success` may put the developer in a false sense of security that their code is doing what they are thinking, and causes them to add extra lines where it may not be needed in some instances. This makes it more clear as to what is happening and doesn't necessarily stop their code from running, but allows them to better debug and bug fix.
- Added the new enums in at lower values than what was already there. Wanted to have them in order of how they appear in `WindowsMicrophoneStream.cs` but didn't want to change said order and end up breaking people's code that may be relying on the actual integer values rather than just the enums. It may not be that likely, but this at least keeps compatibility for those people in those niche cases
- Added new enums to `ErrorCodes` in `MicStreamSelector.cpp` and Booleans to allow for this new error code data to be properly captured by Unity. Different functions will output these new integer values given by the enums and set Booleans to keep track of these states.
- Added XML comments for documentation

DarcyWilson-MU committed:
Fixed Incorrect Error Codes

NOTES:
- Error codes were using old names, or were incorrect. Have been corrected to reflect their current names and to use the error code appropriate the state it is reporting on

DarcyWilson-MU commented:
> @DarcyWilson-MU please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement
> ### Contribution License Agreement
> This Contribution License Agreement (**“Agreement”**) is agreed to by the party signing below (**“You”**), and conveys certain license rights to Microsoft Corporation and its affiliates (“Microsoft”) for Your contributions to Microsoft open source projects. This Agreement is effective as of the latest signature date below.
> 
> 1. **Definitions**.
>    **“Code”** means the computer software code, whether in human-readable or machine-executable form,
>    that is delivered by You to Microsoft under this Agreement.
>    **“Project”** means any of the projects owned or managed by Microsoft and offered under a license
>    approved by the Open Source Initiative ([www.opensource.org](http://www.opensource.org)).
>    **“Submit”** is the act of uploading, submitting, transmitting, or distributing code or other content to any
>    Project, including but not limited to communication on electronic mailing lists, source code control
>    systems, and issue tracking systems that are managed by, or on behalf of, the Project for the purpose of
>    discussing and improving that Project, but excluding communication that is conspicuously marked or
>    otherwise designated in writing by You as “Not a Submission.”
>    **“Submission”** means the Code and any other copyrightable material Submitted by You, including any
>    associated comments and documentation.
> 2. **Your Submission**. You must agree to the terms of this Agreement before making a Submission to any
>    Project. This Agreement covers any and all Submissions that You, now or in the future (except as
>    described in Section 4 below), Submit to any Project.
> 3. **Originality of Work**. You represent that each of Your Submissions is entirely Your original work.
>    Should You wish to Submit materials that are not Your original work, You may Submit them separately
>    to the Project if You (a) retain all copyright and license information that was in the materials as You
>    received them, (b) in the description accompanying Your Submission, include the phrase “Submission
>    containing materials of a third party:” followed by the names of the third party and any licenses or other
>    restrictions of which You are aware, and (c) follow any other instructions in the Project’s written
>    guidelines concerning Submissions.
> 4. **Your Employer**. References to “employer” in this Agreement include Your employer or anyone else
>    for whom You are acting in making Your Submission, e.g. as a contractor, vendor, or agent. If Your
>    Submission is made in the course of Your work for an employer or Your employer has intellectual
>    property rights in Your Submission by contract or applicable law, You must secure permission from Your
>    employer to make the Submission before signing this Agreement. In that case, the term “You” in this
>    Agreement will refer to You and the employer collectively. If You change employers in the future and
>    desire to Submit additional Submissions for the new employer, then You agree to sign a new Agreement
>    and secure permission from the new employer before Submitting those Submissions.
> 5. **Licenses**.
> 
> * **Copyright License**. You grant Microsoft, and those who receive the Submission directly or
>   indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license in the
>   Submission to reproduce, prepare derivative works of, publicly display, publicly perform, and distribute
>   the Submission and such derivative works, and to sublicense any or all of the foregoing rights to third
>   parties.
> * **Patent License**. You grant Microsoft, and those who receive the Submission directly or
>   indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license under
>   Your patent claims that are necessarily infringed by the Submission or the combination of the
>   Submission with the Project to which it was Submitted to make, have made, use, offer to sell, sell and
>   import or otherwise dispose of the Submission alone or with the Project.
> * **Other Rights Reserved**. Each party reserves all rights not expressly granted in this Agreement.
>   No additional licenses or rights whatsoever (including, without limitation, any implied licenses) are
>   granted by implication, exhaustion, estoppel or otherwise.
> 
> 6. **Representations and Warranties**. You represent that You are legally entitled to grant the above
>    licenses. You represent that each of Your Submissions is entirely Your original work (except as You may
>    have disclosed under Section 3). You represent that You have secured permission from Your employer to
>    make the Submission in cases where Your Submission is made in the course of Your work for Your
>    employer or Your employer has intellectual property rights in Your Submission by contract or applicable
>    law. If You are signing this Agreement on behalf of Your employer, You represent and warrant that You
>    have the necessary authority to bind the listed employer to the obligations contained in this Agreement.
>    You are not expected to provide support for Your Submission, unless You choose to do so. UNLESS
>    REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING, AND EXCEPT FOR THE WARRANTIES
>    EXPRESSLY STATED IN SECTIONS 3, 4, AND 6, THE SUBMISSION PROVIDED UNDER THIS AGREEMENT IS
>    PROVIDED WITHOUT WARRANTY OF ANY KIND, INCLUDING, BUT NOT LIMITED TO, ANY WARRANTY OF
>    NONINFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
> 7. **Notice to Microsoft**. You agree to notify Microsoft in writing of any facts or circumstances of which
>    You later become aware that would make Your representations in this Agreement inaccurate in any
>    respect.
> 8. **Information about Submissions**. You agree that contributions to Projects and information about
>    contributions may be maintained indefinitely and disclosed publicly, including Your name and other
>    information that You submit with Your Submission.
> 9. **Governing Law/Jurisdiction**. This Agreement is governed by the laws of the State of Washington, and
>    the parties consent to exclusive jurisdiction and venue in the federal courts sitting in King County,
>    Washington, unless no federal subject matter jurisdiction exists, in which case the parties consent to
>    exclusive jurisdiction and venue in the Superior Court of King County, Washington. The parties waive all
>    defenses of lack of personal jurisdiction and forum non-conveniens.
> 10. **Entire Agreement/Assignment**. This Agreement is the entire agreement between the parties, and
>     supersedes any and all prior agreements, understandings or communications, written or oral, between
>     the parties relating to the subject matter hereof. This Agreement may be assigned by Microsoft.

@microsoft-github-policy-service

keveleigh commented on Input/MicStreamSelector/UnityAddon/WindowsMicrophoneStream.cs:
```diff
@@ -99,10 +99,12 @@ public WindowsMicrophoneStreamErrorCode Initialize(WindowsMicrophoneStreamType s
             if (initialized)
             {
                 // The microphone stream is already initialized, no need to alarm the calling code.
```
nit: should we edit these comments throughout, now that we aren't returning `Success`?

DarcyWilson-MU replied:
Not a problem! Appreciate you taking a look, and no stress if no-one see's it, I know it's been out of active development for a while, but I was fixing some things on my end for a project and thought I may PR it just in case, haha. Still on MRTK2 for the time being.

Can definitely chuck those comments out though, wasn't sure whether to keep them or toss them as I didn't want to muck around with the codebase too much. Will make another edit in just a moe.

keveleigh reviewed (commented):
Looks good to me! Though, I'm not sure anybody is monitoring this repo at Microsoft anymore 😢 I can ping some folks and see

DarcyWilson-MU committed:
Removed + Updated Comments

NOTES:
- Removed the old comments from the if-statement checks in each of the methods that had them as they are no longer just returning `Success` error codes
- Updated the XML comments for the methods with the if-statement checks that had their comments removed. Noted that the `WindowsMicrophoneStreamErrorCode` can either indicate success, failure, or a reason why it did not progress past the if-statement check
